gordon_janitor_gcp/plugins/authority.py

```diff
@@ -0,0 +1,185 @@
+                             blacklisted_tags=tag_blacklist,
+                             blacklisted_metadata=metadata_blacklist)
+
+    def build_authority(self):
+        """Coordinate dependencies and create a GCEAuthority object.
+
+        Returns:
+            obj: A configured GCEAuthority object.
+        """
+        self._get_or_raise('zone')
+        keyfile_path = self._get_or_raise('keyfile')
+        scopes = self._get_or_raise('scopes')
+        crm_client = self._get_crm_client(keyfile_path, scopes)
+        gce_client = self._get_gce_client(keyfile_path, scopes)
+
+        return GCEAuthority(self.config, crm_client, gce_client,
+                            self.rrset_channel, **self.kwargs)
+
+
+@zope.interface.implementer(interfaces.IAuthority)
+class GCEAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
+
+    Args:
+        config (dict): plugin-specific configuration.
+        crm_client (obj): client used to fetch GCE projects.
+        gce_client (obj): client used to fetch instances for each project.
+        rrset_channel (asyncio.Queue): channel to send resource record messages
+            to.
+    """
+
+    def __init__(self, config, crm_client, gce_client, rrset_channel=None,
+                 **kwargs):
+        self.config = config
+        self.crm_client = crm_client
+        self.gce_client = gce_client
+        self.rrset_channel = rrset_channel
+
+    async def _get_active_project_ids(self):
+        active_projects = await self.crm_client.list_all_active_projects()
+        return set(p.get('projectId') for p in active_projects)
+
+    async def _get_projects(self):
+        projects = await self._get_active_project_ids()
+        project_blacklist = set(self.config.get('project_blacklist', []))
+        # TODO: emit a metric for all projects and a metric for
+        # project - blacklist.
+        return sorted(projects - project_blacklist)
+
+    async def _get_instances(self, projects):
+        instance_filter = self.config.get('instance_filter')
+        for project in projects:
+            yield await self.gce_client.list_instances(
+                project, instance_filter=instance_filter)
+
+    def _create_instance_rrset(self, instance):
+        return {
+            'name': instance['hostname'],
+            'type': 'A',
+            'rrdatas': [instance['external_ip']]
+        }
+
+    def _create_msg(self, instances):
+        return {
+            'zone': self.config['zone'],
+            'rrsets': [
+                self._create_instance_rrset(instance_data)
+                for instance_data in instances
+            ]
+        }
+
+    async def start(self):
+        """Batch instance data and send it to the :py:obj:`self.rrset_channel`.
+        """
+        projects = await self._get_projects()
+
+        instances = []
+        async for project_instances in self._get_instances(projects):
+            instances.extend(project_instances)
```

Comment:
@econchick I used `extend` here because it's a bit clearer. I also looked into what you said, about `extend` being less efficient, but I couldn't find anything about this except the Python [TimeComplexity - list](https://wiki.python.org/moin/TimeComplexity) page, which lists that single appends are O(1) whereas extends are O(k), which makes it sound like they should be similarly efficient.